pkg/controller/servingcert/controller/secret_updating_controller.go

```diff
@@ -69,70 +70,105 @@ func (sc *serviceServingCertUpdateController) Key(namespace, name string) (metav
+			update = true
+		}
+	}
+	if !update {
+		return nil
+	}
+	_, err = sc.secretClient.Secrets(secretCopy.Namespace).Update(secretCopy)
 	return err
 }
 
-func (sc *serviceServingCertUpdateController) requiresRegeneration(secret *v1.Secret) (bool, *v1.Service) {
-	serviceName, ok := toServiceName(secret)
-	if !ok {
-		return false, nil
+func (sc *serviceServingCertUpdateController) isSecretValidForService(sharedService *v1.Service, secret *v1.Secret) bool {
+	isValid := true
+	if sharedService.Annotations[api.ServingCertSecretAnnotation] != secret.Name && sharedService.Annotations[api.AlphaServingCertSecretAnnotation] != secret.Name {
+		isValid = false
 	}
+	if secret.Annotations[api.ServiceUIDAnnotation] != string(sharedService.UID) && secret.Annotations[api.AlphaServiceUIDAnnotation] != string(sharedService.UID) {
+		isValid = false
+	}
+	return isValid
+}
 
-	sharedService, err := sc.serviceLister.Services(secret.Namespace).Get(serviceName)
+func (sc *serviceServingCertUpdateController) getServiceForSecret(sharedSecret *v1.Secret) *v1.Service {
+	serviceName, ok := toServiceName(sharedSecret)
+	if !ok {
+		return nil
+	}
+	service, err := sc.serviceLister.Services(sharedSecret.Namespace).Get(serviceName)
 	if kapierrors.IsNotFound(err) {
-		return false, nil
+		return nil
 	}
 	if err != nil {
-		utilruntime.HandleError(fmt.Errorf("unable to get service %s/%s: %v", secret.Namespace, serviceName, err))
-		return false, nil
-	}
-
-	if sharedService.Annotations[api.ServingCertSecretAnnotation] != secret.Name && sharedService.Annotations[api.AlphaServingCertSecretAnnotation] != secret.Name {
-		return false, nil
-	}
-	if secret.Annotations[api.ServiceUIDAnnotation] != string(sharedService.UID) && secret.Annotations[api.AlphaServiceUIDAnnotation] != string(sharedService.UID) {
-		return false, nil
+		utilruntime.HandleError(fmt.Errorf("unable to get service %s/%s: %v", sharedSecret.Namespace, serviceName, err))
+		return nil
 	}
+	return service
+}
 
+func (sc *serviceServingCertUpdateController) requiresRegeneration(service *v1.Service, secret *v1.Secret, minTimeLeft time.Duration) bool {
 	// if we don't have an ownerref, just go ahead and regenerate.  It's easier than writing a
 	// secondary logic flow.
-	if !ocontroller.HasOwnerRef(secret, ownerRef(sharedService)) {
-		return true, sharedService
+	if !ocontroller.HasOwnerRef(secret, ownerRef(service)) {
+		return true
 	}
-
 	// if we don't have the annotation for expiry, just go ahead and regenerate.  It's easier than writing a
 	// secondary logic flow that creates the expiry dates
 	expiryString, ok := secret.Annotations[api.ServingCertExpiryAnnotation]
 	if !ok {
 		expiryString, ok = secret.Annotations[api.AlphaServingCertExpiryAnnotation]
 		if !ok {
-			return true, sharedService
+			return true
 		}
 	}
 	expiry, err := time.Parse(time.RFC3339, expiryString)
 	if err != nil {
-		return true, sharedService
+		return true
 	}
 
 	if time.Now().Add(sc.minTimeLeftForCert).After(expiry) {
```

Comment:
Use the sc.minTimeLeftForCert or the minTimeLeft argument (using the struct member instead of an extra argument is fine)